// bybit-api/lib/util/rest/fileUploadUtil.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getMimeType = getMimeType;
exports.buildMultipartPayload = buildMultipartPayload;
/**
 * Assume MIME type from filename extension
 */
function getMimeType(filename) {
    var _a;
    const ext = (_a = filename.split('.').pop()) === null || _a === void 0 ? void 0 : _a.toLowerCase();
    const mimeTypes = {
        jpg: 'image/jpeg',
        jpeg: 'image/jpeg',
        png: 'image/png',
        pdf: 'application/pdf',
        mp4: 'video/mp4',
    };
    return mimeTypes[ext || ''] || 'application/octet-stream';
}
/**
 * @private Build multipart/form-data payload for file uploads (Node.js only)
 */
function buildMultipartPayload(params) {
    return __awaiter(this, void 0, void 0, function* () {
        const fileBuffer = params.fileBuffer;
        const fileName = params.fileName;
        if (!fileBuffer) {
            throw new Error('upload_file parameter is required for file uploads');
        }
        if (!fileName) {
            throw new Error('filename parameter is required for file uploads (used for MIME type detection)');
        }
        if (!Buffer.isBuffer(fileBuffer)) {
            throw new Error('uploadFile must be provided as a Buffer. To upload from a file path, read the file yourself: fs.readFileSync(filePath)');
        }
        const fileData = { data: fileBuffer, filename: fileName };
        // Determine MIME type from filename extension
        const mimeType = getMimeType(fileData.filename);
        // Build multipart/form-data manually (following Bybit's format)
        const boundary = 'boundary-for-file';
        const contentType = `multipart/form-data; boundary=${boundary}`;
        const header = `--${boundary}\r\nContent-Disposition: form-data; name="upload_file"; filename="${fileData.filename}"\r\nContent-Type: ${mimeType}\r\n\r\n`;
        const footer = `\r\n--${boundary}--\r\n`;
        // Construct payload (Node.js only - simple Buffer concatenation)
        const payload = Buffer.concat([
            Buffer.from(header, 'utf-8'),
            fileData.data,
            Buffer.from(footer, 'utf-8'),
        ]);
        return { payload, contentType };
    });
}
